internal/scheduler/common.go

```diff
@@ -0,0 +1,278 @@
+package scheduler
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	"golang.org/x/exp/maps"
+
+	"github.com/armadaproject/armada/internal/armada/configuration"
+	armadamaps "github.com/armadaproject/armada/internal/common/maps"
+	armadaresource "github.com/armadaproject/armada/internal/common/resource"
+	armadaslices "github.com/armadaproject/armada/internal/common/slices"
+	schedulerconfig "github.com/armadaproject/armada/internal/scheduler/configuration"
+	schedulercontext "github.com/armadaproject/armada/internal/scheduler/context"
+	"github.com/armadaproject/armada/internal/scheduler/interfaces"
+	"github.com/armadaproject/armada/internal/scheduler/schedulerobjects"
+)
+
+// SchedulerResult is returned by Rescheduler.Schedule().
+type SchedulerResult struct {
+	// Running jobs that should be preempted.
+	PreemptedJobs []interfaces.LegacySchedulerJob
+	// Queued jobs that should be scheduled.
+	ScheduledJobs []interfaces.LegacySchedulerJob
+	// For each preempted job, maps the job id to the id of the node on which the job was running.
+	// For each scheduled job, maps the job id to the id of the node on which the job should be scheduled.
+	NodeIdByJobId map[string]string
+}
+
+func NewSchedulerResult[S ~[]T, T interfaces.LegacySchedulerJob](
+	preemptedJobs S,
+	scheduledJobs S,
+	nodeIdByJobId map[string]string,
+) *SchedulerResult {
+	castPreemptedJobs := make([]interfaces.LegacySchedulerJob, len(preemptedJobs))
+	for i, job := range preemptedJobs {
+		castPreemptedJobs[i] = job
+	}
+	castScheduledJobs := make([]interfaces.LegacySchedulerJob, len(scheduledJobs))
+	for i, job := range scheduledJobs {
+		castScheduledJobs[i] = job
+	}
+	return &SchedulerResult{
+		PreemptedJobs: castPreemptedJobs,
+		ScheduledJobs: castScheduledJobs,
+		NodeIdByJobId: nodeIdByJobId,
+	}
+}
+
+// PreemptedJobsFromSchedulerResult returns the slice of preempted jobs in the result,
+// cast to type T.
+func PreemptedJobsFromSchedulerResult[T interfaces.LegacySchedulerJob](sr *SchedulerResult) []T {
+	rv := make([]T, len(sr.PreemptedJobs))
+	for i, job := range sr.PreemptedJobs {
+		rv[i] = job.(T)
+	}
+	return rv
+}
+
+// ScheduledJobsFromScheduleResult returns the slice of scheduled jobs in the result,
+// cast to type T.
+func ScheduledJobsFromSchedulerResult[T interfaces.LegacySchedulerJob](sr *SchedulerResult) []T {
+	rv := make([]T, len(sr.ScheduledJobs))
+	for i, job := range sr.ScheduledJobs {
+		rv[i] = job.(T)
+	}
+	return rv
+}
+
+func JobsSummary(jobs []interfaces.LegacySchedulerJob) string {
```

Comment:
This function probably warrents a comment about what it does, along with a perfromance warning as I suspect it could be quite slow if `jobs` is large.

Reply:
I'd also ague that this function sounds incredible generic ("scheduler.JobsSummary") but to me looks like it's producing a string fromatted for relatively specific use case.  E.g. it straight away makes the assuption that the jobs passed in have all been evicted.